AxonDeepSeg/train_network.py

```diff
@@ -75,63 +73,104 @@ def train_model(path_trainingset, path_model, config, path_model_init=None,
+    flipping = config["da-4-flipping-activate"]
+    gaussian_blur = config["da-5-gaussian_blur-activate"]
+    reflection_border = config["da-6-reflection_border-activate"] #Config parameter to determine whether relection or constant(value = 0) is used for border pixel values while performing augmentation operations such as rotation, rescaling and shifting.
+
+    if reflection_border:
+        border_mode = cv2.BORDER_REFLECT_101
+    else:
+        border_mode = cv2.BORDER_CONSTANT
+
+    p_shift = p_rescale = p_rotate = p_elastic = p_flip = p_blur = 0
+    #If the key values of augmentation are set to True then their respective probability are set to 0.5 else to 0. Probalility(p) suggests a certainity of applying data augmentation operations (shift, rotate, blur, elastic, flip) to an image.
+
+    # Probability value of 0.5 is chosen so that the original as well augmented image are taken into account while training the model.
+    if shifting:
+        p_shift = 0.5
+    if rotation:
+        p_rotate = 0.5
+    if flipping:
+        p_flip = 0.5
+    if gaussian_blur:
+        p_blur = 0.5
+    if elastic:
+        p_elastic = 0.5
+
+
+    #####Data Augmentation parameters#####
+
+    #Elastic transform parameters
+    alpha_max = 9
+    sigma = 3
+    alpha = random.choice(list(range(1, alpha_max)))
+
+    #Random rotation parameters
+    low_bound = 5
+    high_bound = 89
+
+    #Shifting parameters
+    percentage_max = 0.1
+    size_shift = int(percentage_max*image_size)
+    low_limit = 0
+    high_limit = (2*size_shift-1) / image_size
+
 
-    #####################Training###################
-    image_datagen = ImageDataGenerator(**data_gen_args)
-    mask_datagen = ImageDataGenerator(**data_gen_args)
+    ######################################
 
-    seed = 2018
 
-    # Image and Mask Data Generator
-    image_generator_train = image_datagen.flow(image_train_gen, y=None, seed = seed,  batch_size=batch_size, shuffle=True)
-    mask_generator_train = mask_datagen.flow(mask_train_gen, y=None, seed = seed, batch_size=batch_size, shuffle=True)
+    AUGMENTATIONS_TRAIN = Compose([
 
-    # Just zip the two generators to get a generator that provides augmented images and masks at the same time
-    train_generator = zip(image_generator_train, mask_generator_train)
+    #Randomy flips an image either horizontally, vertically or both.
+    Flip(p = p_flip),
 
-    ###########################Validation###########
-    data_gen_args = dict()
-    image_datagen = ImageDataGenerator(**data_gen_args)
-    mask_datagen = ImageDataGenerator(**data_gen_args)
+    #Randomly rotates an image between low limit and high limit.
+    ShiftScaleRotate(shift_limit=(low_limit, high_limit) , scale_limit=(0,0),rotate_limit=(0,0), border_mode=border_mode, p = p_shift, interpolation = cv2.INTER_NEAREST),
 
+    #Randomly applies elastic transformation on the image.
+    ElasticTransform(alpha= alpha, sigma= sigma, p = p_elastic, alpha_affine = alpha),
```

Comment:
Can the ElasticTransform not have nearest neighbour interpolation for the masks with Augmentations? If not, we should maybe explore omitting this augmentation, since the masks will have more than 3 classes. Otherwise, we might need to rebin the mask to the three classes after this augmentation is performed if possible.

Reply:
Elastic transform and GaussianBlur don't have nearest neighbour interpolation for masks. Could you elaborate on `omitting the augmentation` as it's unclear to me ?

Reply:
This is a point that needs to be revisited, as it wasn't resolved. By not having nearest neighbour intperpolation, elastic and gaussianblur transformations will cause data to be trained on masks with more than three classes if used .

Reply:
But in the case of augmentation applied with Elastic and Gaussian Blur, i didn't see the behaviour of more than 3 classes. I am not sure that it impact our results.

Reply:
The final output masks won't have more than 3 classes, because we likely thresholded them. We need to verify if the intermediate data (i.e. augmented dataset generated by these functions) have more than three classes. If you're telling me that they don't have a "nearest neighbour" function, then it's very likely that they'll be interpolating between the values (i.e. between 255 and 127 for the axon/myelin border), meaning more than 3 classes being used to train the data.

Reply:
Agreed and  I was wrong for the` ElasticTranform`, they do have interpolation mode of nearest neighbour( earlier, I must have seen the parameter of other augmentation operation on the documentation)

Reply:
That's great news! Do they have it for Gaussian as well?

Reply:
I don't think there is a need to Gaussian as it just add noise to the data. And they don't have for Gaussian.

Reply:
Resolved Elastic Transform interpolation to Nearest Neighbour in commit(dae4a1d).